src/module_3/module_3.md

```diff
@@ -0,0 +1,891 @@
+
+oversample = SMOTE()
+X_smote, y_smote = oversample.fit_resample(X_train_scaled, y_train)
+
+
+
+models = [('SGD LOG', sgd_log),
+          ('SGD SVM', sgd_svm),
+          ('Ridge', ridge),
+          ('Logistic Regression', log_reg)]
+
+accuracy = []
+precision = []
+recall = []
+names = []
+
+
+for name, model in models:
+    scores = cross_validate(model, X_smote, y_smote, cv=10, scoring=['accuracy', 'precision', 'recall'])
+    accuracy.append(scores['test_accuracy'].mean())
+    precision.append(scores['test_precision'].mean())
+    recall.append(scores['test_recall'].mean())
+    names.append(name)
+
+
+results = pd.DataFrame({'Model': names, 'Accuracy': accuracy, 'Precision': precision, 'Recall': recall})
+
+results
+```
+
+
+
+
+<div>
+<style scoped>
+    .dataframe tbody tr th:only-of-type {
+        vertical-align: middle;
+    }
+
+    .dataframe tbody tr th {
+        vertical-align: top;
+    }
+
+    .dataframe thead th {
+        text-align: right;
+    }
+</style>
+<table border="1" class="dataframe">
+  <thead>
+    <tr style="text-align: right;">
+      <th></th>
+      <th>Model</th>
+      <th>Accuracy</th>
+      <th>Precision</th>
+      <th>Recall</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th>0</th>
+      <td>SGD LOG</td>
+      <td>0.739644</td>
+      <td>0.800940</td>
+      <td>0.637845</td>
+    </tr>
+    <tr>
+      <th>1</th>
+      <td>SGD SVM</td>
+      <td>0.736903</td>
+      <td>0.807866</td>
+      <td>0.621708</td>
+    </tr>
+    <tr>
+      <th>2</th>
+      <td>Ridge</td>
+      <td>0.712254</td>
+      <td>0.804072</td>
+      <td>0.561273</td>
+    </tr>
+    <tr>
+      <th>3</th>
+      <td>Logistic Regression</td>
+      <td>0.739791</td>
+      <td>0.800841</td>
+      <td>0.638327</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+</p>As whe have class imbalance, we use `class_weight='balance'` and oversample to increase the number of samples in the underrepresented class.</p>
+<p>Based on the previous cells, we stick with Logistic Regression that has a higher score across accuracy, precision and recall. Although, due to the class imbalance the metric we are most interested in is recall.</p>
+<p>Let's now fine-tune its hyperparameters.</p>
+
+### Final model
+
+
+```python
+from sklearn.metrics import confusion_matrix
+
+params = {
+    'C': [0.001, 0.01, 0.1, 1, 10, 100]
+    }
+
+log_reg = LogisticRegression(class_weight='balanced')
+
+grid = GridSearchCV(log_reg, cv=10, param_grid=params, n_jobs=-1, refit=True, scoring='precision')
+grid.fit(X_smote, y_smote) 
+print(grid.best_params_)
+print(grid.best_score_)
+print(grid.best_estimator_)
+y_pred = grid.predict(X_test_scaled)
+y_probs = grid.predict_proba(X_test_scaled)[:,1]
+y_log_probs = grid.predict_log_proba(X_test_scaled)[:,1]
+print(classification_report(y_test, y_pred))
+print(confusion_matrix(y_test, y_pred))
+```
+
+    {'C': 0.001}
+    0.800954386060868
+    LogisticRegression(C=0.001, class_weight='balanced')
```

Comment:
Again, you are using a concrete threshold when probably the business wants to choose a concrete tradeoff of precision / recall among all of them